File: golang/config/self.go
package config

import (
	"regexp"
	"runtime"
	"strings"
)

var (
	gitCommitHash = "unknown commit"
	buildDate     = "unknown time" // UTC time "yyyy-MM-ddTHH:mm:ssZ"

	regex = regexp.MustCompile(`^[\d.]+$`)
)

type h map[string]interface{}

var MetaInfo = h{
	"name":    Name,
	"version": Version,
	"dev":     !regex.MatchString(Version), // development version or stable release
	"info": h{
		"homepage":    Homepage,
		"description": LongDescription,
	},
	"meta": h{
		"implementation": h{
			"platform": runtime.GOOS + "/" + runtime.GOARCH,
			"tool":     "golang",
			"version":  strings.TrimPrefix(runtime.Version(), "go"),
		},
		"git_commit_hash": gitCommitHash,
		"build_at":        buildDate,
	},
	"options": h{
		"instruction": "https://github.com/LussacZheng/dandanplay-resource-service/tree/main/docs",
		"supported":   []string{"$realtime", "$page", "$limit"},
	},
}

// https://github.com/gin-gonic/gin/issues/2360

File: typescript/packages/dandanplay-resource-api/src/self.ts
type MetaInfo = {
  name: string
  version: string
  dev: boolean
  info: {
    homepage: string
    description: string
  }
  meta: {
    implementation: {
      platform: string
      tool: string
      version: string
    }
    git_commit_hash: string
    build_at: string
  }
  options: {
    instruction: string
    supported: string[]
  }
}

type CoreInfo = 'name' | 'version' | 'homepage' | 'description' | 'platform' | 'tool'

export function generateMetaInfo(coreInfo: Record<CoreInfo, string>): MetaInfo {
  return {
    name: coreInfo.name,
    version: coreInfo.version,
    dev: !/^[\d\.]+$/.test(coreInfo.version),
    info: {
      homepage: coreInfo.homepage,
      description: coreInfo.description,
    },
    meta: {
      implementation: {
        platform: coreInfo.platform,
        tool: coreInfo.tool,
        version: '${toolVersion}',
      },
      git_commit_hash: '${gitCommitHash}',
      build_at: '${buildDate}',
    },
    options: {
      instruction: 'https://github.com/LussacZheng/dandanplay-resource-service/tree/main/docs',
      supported: ['$realtime', '$page', '$limit'],
    },
  }
}
